e3nn/networks.py

```diff
@@ -132,6 +132,62 @@ def forward(self, input, *args, **kwargs):
         return output
 
 
+class S2ConvNetwork(torch.nn.Module):
+    def __init__(self, Rs_in, mul, Rs_out, lmax, layers=3,
+                 max_radius=1.0, number_of_basis=3, radial_layers=3,
+                 kernel=Kernel, convolution=Convolution):
+        super().__init__()
+
+        def parity(l):
+            if l % 2 == 0:
+                return 1
+            else:
+                return -1
+
+        Rs_hidden = [(1, l, (-1)**l) for i in range(mul) for l in range(lmax + 1)]
+        representations = [Rs_in]
+        representations += [Rs_hidden] * layers
+        representations += [Rs_out]
+
+        RadialModel = partial(GaussianRadialModel, max_radius=max_radius,
+                              number_of_basis=number_of_basis, h=100,
+                              L=radial_layers, act=swish)
+
+        K = partial(kernel, RadialModel=RadialModel, selection_rule=partial(o3.selection_rule_in_out_sh, lmax=lmax))
+
+        def make_layer(Rs_in, Rs_out):
+            act = S2Activation([(1, l, parity(l)) for l in range(lmax + 1)], sigmoid, lmax_out=lmax, res=20 * (lmax + 1))
+            conv = convolution(K(Rs_in, Rs_out))
+            return torch.nn.ModuleList([conv, act])
+
+        self.layers = torch.nn.ModuleList([
+            make_layer(Rs_layer_in, Rs_layer_out)
+            for Rs_layer_in, Rs_layer_out in zip(representations[:-2], representations[1:-1])
+        ])
+
+        self.layers.append(convolution(K(representations[-2], representations[-1])))
+        self.mul = mul
+        self.lmax = lmax
+
+    def forward(self, input, *args, **kwargs):
+        output = input
+        N = args[0].shape[-2]
```

Comment:
if `args[0]` is known to be a tensor why don't you name it?

Reply:
because what it is depends on the Convolution used. It's either the geometry for the standard point.operations.Convolution or it's the edge_index for point.message_passing.Convolution.